app/controllers/moderator_applications_controller.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+class ModeratorApplicationsController < ApplicationController
+  before_action :set_application, only: %i[show edit update destroy]
+  before_action :ensure_authenticated, only: %i[index show new create edit update destroy]
+  before_action :ensure_authorized_or_admin, only: %i[show edit update destroy]
+  before_action :ensure_applications_open, only: %i[show new create edit update destroy]
+  before_action :prevent_conflict, only: %i[new create]
+
+  def index
+    # Only the administrator can see moderator applications other than their own.
+    unless current_user.is_admin
+      if current_user.moderator_application
+        redirect_to current_user.moderator_application
+      else
+        redirect_to new_moderator_application_path
+      end
+      return
+    end
+
+    @applications = ModeratorApplication.all.order(created_at: :desc)
+  end
+
+  def show; end
+
+  def new
+    @application = ModeratorApplication.new
+  end
+
+  def create
+    @application = ModeratorApplication.new(application_params)
+    @application.user_id = current_user.id
+
+    view_result @application.save
+  end
+
+  def edit; end
+
+  def update
+    view_result @application.update(application_params)
+  end
+
+  def destroy
+    @application.destroy!
+
+    if current_user.is_admin
+      redirect_to moderator_applications_path
+    else
+      redirect_to root_url
+    end
+  end
+
+  private
+
+  def set_application
+    @application = ModeratorApplication.find_by(id: params[:id])
+
+    render_not_found if @application.nil?
+  end
+
+  def application_params
+    params.require(:moderator_application)
```

Comment:
Right now, the controller crashes if you don't pass a `:moderator_application` parameter, but when trying to figure out how to solve it, it almost looked like rails is designed so that uncaught exceptions are a standard error handling mechanism? I hate it, but what am I supposed to do? `try/catch` this and just render an error like Rails would have anyway? Not to mention this doesn't even do any general-purpose input validation; that's done by the model.

Implicit-exceptions-as-control-flow seems *wrong*, but I think that might just be standard Rails practice. (And don't get me wrong; I'm a big fan of continuations/effects (*not* exceptions) for control flow, but I like them to be *explicit*, which Ruby inherently isn't.)

Reply:
There's also the built in
```ruby
rescue_from  ActionController::ParameterMissing, :with => :render_whatever
```
which seems reasonable because it's at least built-in, kinda like the `before_action` callbacks are? I still don't like it, but at least it's standard practice. It could also probably be put on the base `ApplicationController`. Maybe throw back a `status: :unprocessable_entity` in that case? Or maybe a redirect. Though it's not like this issue is ever going to matter anyway!

Reply:
What possible case could arise that result in :moderator_application not being passed?

Reply:
In practice? Nothing. You'd almost have to deliberately fuck it up. I wasn't even planning on fixing it when I was in that weird mindset; I was just speculating about what the best control flow mechanism would be.